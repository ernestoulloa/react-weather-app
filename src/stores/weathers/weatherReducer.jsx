import ACTION_TYPES from './weatherActionTypes';

export const initialState = {
  isLoading: true,
  errors: null,
  cod: '',
  message: 0,
  cnt: 0,
  list: '',
  city: '',
};

function weatherReducer(state = initialState, action = {}) {
  switch (action.type) {
    case ACTION_TYPES.WEATHERS_GET_ALL_REQUEST: {
      return {
        ...state,
        errors: false,
        isLoading: true,
      };
    }

    case ACTION_TYPES.WEATHERS_GET_ALL_SUCCESS: {
      return {
        ...state,
        isLoading: false,
        errors: false,
        cod: action.payload.cod,
        message: action.payload.message,
        cnt: action.payload.cnt,
        list: action.payload.list,
        city: action.payload.city,
      };
    }

    case ACTION_TYPES.WEATHERS_GET_ALL_FAILURE: {
      return {
        ...state,
        errors: true,
        isLoading: false,
      };
    }

    default:
      return state;
  }
}

export default weatherReducer;
