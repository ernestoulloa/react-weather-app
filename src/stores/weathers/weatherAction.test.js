import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import weatherReducer, { initialState } from './weatherReducer';
import ACTION_TYPES from './weatherActionTypes';
import getAll from './weatherAction';
import API from '../../services';
import { mockResponse } from '../../constants/constants';

describe('ACTION_TYPES.WEATHERS_GET_ALL_REQUEST', () => {
  it('should return the initial state', () => {
    expect(weatherReducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle ACTION_TYPES.WEATHERS_GET_ALL_REQUEST', () => {
    expect(
      weatherReducer(initialState,
        {
          type: 'ACTION_TYPES.WEATHERS_GET_ALL_REQUEST',
        }),
    ).toMatchSnapshot();
  });
});

describe('Get Forecast data', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });

  const fetchForecastData = async () => API.weathers.getAll().then((res) => res.json());

  it('Fetches the forecast data', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockResponse));
    const response = await fetchForecastData();
    expect(response.list).toBeDefined();
    expect(fetch).toHaveBeenCalledTimes(1);
  });

  it('dispatches proper actions', () => {
    const middlewares = [thunk];
    const mockStore = configureMockStore(middlewares);
    const store = mockStore(initialState);
    fetch.mockResponseOnce(JSON.stringify(mockResponse));

    return store.dispatch(getAll()).then(() => {
      const actions = store.getActions();
      expect(actions[0].type).toEqual(ACTION_TYPES.WEATHERS_GET_ALL_REQUEST);
      expect(actions[1].type).toEqual(ACTION_TYPES.WEATHERS_GET_ALL_SUCCESS);
    });
  });
});
