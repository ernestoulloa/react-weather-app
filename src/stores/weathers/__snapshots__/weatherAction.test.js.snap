// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ACTION_TYPES.WEATHERS_GET_ALL_REQUEST should handle ACTION_TYPES.WEATHERS_GET_ALL_REQUEST 1`] = `
Object {
  "city": "",
  "cnt": 0,
  "cod": "",
  "errors": null,
  "isLoading": true,
  "list": "",
  "message": 0,
}
`;

exports[`ACTION_TYPES.WEATHERS_GET_ALL_REQUEST should return the initial state 1`] = `
Object {
  "city": "",
  "cnt": 0,
  "cod": "",
  "errors": null,
  "isLoading": true,
  "list": "",
  "message": 0,
}
`;
